pkg/cmd/server/defaults.go

```diff
@@ -319,10 +306,16 @@ func DefaultUnaryMiddleware(opts MiddlewareOption) (*MiddlewareChain[grpc.UnaryS
 			WithInterceptor(grpcMetricsUnaryInterceptor).
 			Done(),
 
+		NewUnaryMiddleware().
+			WithName(DefaultMiddlewareMemoryProtection).
+			WithInterceptor(memoryProtectionUnaryInterceptor.UnaryServerInterceptor()).
+			Done(),
+
 		NewUnaryMiddleware().
 			WithName(DefaultMiddlewareGRPCAuth).
 			WithInterceptor(grpcauth.UnaryServerInterceptor(opts.AuthFunc)).
-			EnsureAlreadyExecuted(DefaultMiddlewareGRPCProm). // so that prom middleware reports auth failures
+			EnsureAlreadyExecuted(DefaultMiddlewareGRPCProm).         // so that prom middleware reports auth failures
+			EnsureAlreadyExecuted(DefaultMiddlewareMemoryProtection). // so that memory protection happens before auth
```

Comment:
Why does it have to run before auth?

The rationale for `EnsureAlreadyExecuted` is to declare a dependency between middlewares explicitly: "I need this to run before me because I depend on its side effects". So why does the auth middleware depend on memory protection?

If not, you can remove it.

If anything, I'd argue that memory protection depends on all the previous middlewares, because we want observability to take place _before we kick memory protection_, otherwise we would be running blind when `ResourceExhausted` is returned.